lib/msgServer/mmrp/index.js

```diff
@@ -130,28 +147,64 @@ exports.MmrpNode = MmrpNode;
  * @param {Function} [cb]
  */
 
-MmrpNode.prototype.connect = function (uri, clusterId, cb) {
+MmrpNode.prototype.connect = function (uri, data, cb) {
+	const clusterId = data.clusterId;
+	const timestamp = data.timestamp;
+
 	// we cannot connect through routers (long routes), we are establishing a direct connection
 
 	assert.strictEqual(typeof uri, 'string');
 	assert.strictEqual(typeof clusterId, 'string');
 
-	var route = [clusterId];
-	var relay = this.relays[clusterId];
+	// check if other relay(s) with the same URI exist, and disconnect them;
+	// this is required to clean up old relays that have yet to be removed
+	// from the relay list
+
+	for (const relayClusterId in this.relays) {
+		const relay = this.relays[relayClusterId];
+
+		// Filter out our own relay
+		if (relayClusterId === clusterId) {
+			continue;
+		}
+
+		// Filter out relays with a different URI
+		if (relay.uri !== uri) {
+			continue;
+		}
+
+		// If we find a more recent node in our
+		// list, we should then ignore adding
+		// the node we have just received overall
+		if (relay.timestamp >= timestamp) {
+			return callMeMaybe(cb);
+		}
+
+		logger.verbose.data({
+			oldRelayId: relayClusterId,
+			newRelayId: clusterId
+		}).log(this.clusterId, 'found relay with same uri, removing the relay');
+
+		// We delete the relay reference, but do not disconnect; quick disconnect-reconnect
+		// appear to confuse ZeroMQ, and may permanently break the connection in some cases.
+		delete this.relays[relayClusterId];
+	}
+
+	// create route, and look up the relay map
+
+	const route = [clusterId];
+	const relay = this.relays[clusterId];
 
 	// if we're already connected to this URI, we ignore this request
 
 	if (relay && relay.uri) {
-		if (cb) {
-			cb();
-		}
-		return;
+		return callMeMaybe(cb);
```

Comment:
Haven't you ruled this out by the loop you introduce above? Either we've returned because the existing relay is newer, or we've deleted the existing one.

Reply:
I think that's incorrect:

1. We go through the list of nodes
2. We continue if this node is the node we are connecting to
3. We continue if the URIs do not match
4. If we find a more recent node, we return, since the node we are trying to connect is outdated
5. We disconnect any node reaching this point

Therefore, once we reach that point, it's entirely possible that we have disconnected a bunch of servers but were not connected ourselves.